### ISSUE TITLE
JSR 305 @Nonnull does not behave the same way as lombok's @NonNull

### ISSUE THREAD

**Issue by janeisklar** [rzwitserloot/lombok#1943]
Reading the comment on the `lombok.NonNull` annotation, lead me to believe that I could use Findbugs JSR-305 `@Nonnull` annotation as a replacement for lombok's own annotation:

>  Note that any annotation named {@code NonNull} with any casing and any package will result in nullchecks produced for generated methods (and the annotation will be copied to the getter return type and any parameters of generated methods), but <em>only</em> this annotation, if present on a parameter, will result in a null check inserted into your otherwise handwritten method.
> (see https://github.com/rzwitserloot/lombok/blob/master/src/core/lombok/NonNull.java#L35-L38)

However, the following class:

```
import javax.annotation.Nonnull;

import lombok.NonNull;

public class Test {

    public void testFoo(@NonNull final String foo) {
        foo.length();
    }

    public void testFoo2(@Nonnull final String foo2) {
        foo2.length();
    }
}
```

compiles to this:
```
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

import javax.annotation.Nonnull;
import lombok.NonNull;

public class Test {
  public Test() {
  }

  public void testFoo(@NonNull String foo) {
    if (foo == null) {
      throw new NullPointerException("foo");
    } else {
      foo.length();
    }
  }

  public void testFoo2(@Nonnull String foo2) {
    foo2.length();
  }
}
```

Is this a bug or did I simply misread the annotation's comment in some way?

**Comment by janeisklar** [comment]
After re-reading the comment I just noticed that the replacement is intended to work only for field annotations. Is that correct? Is there any way I could configure it such that it would also work for method arguments?

**Comment by rzwitserloot** [comment]
[A] I guess the docs aren't too clear but the behaviour of lombok matches what the docs say: Lombok has 2 utterly unrelated nullity-related things it does: When lombok generates an entire method/constructor (Specifically, `@Setter`, and `@XArgsConstructor`), any well-known 'non-null' annotation triggers lombok to put a null check into the generated code, and, the unrelated thing: if you annotate a method parameter with specifically the `lombok.NonNull` annotation (which by pure coincidence so happens to be one of the 'well known non-null indicating annotations' for the previous thing), we inject a null check right into your hand written method.

So, for your specific bug report: 'nope, works as specified and designed'.

But, there are 2 tangential issues I discovered whilst investigating this report:

1. The docs are still wrong but not in the way you report: Since recently, the list of annotations that trigger lombok's 'oh, I guess if I generate an entire setter/constructor I should add null checks' is no longer 'anything whose simple name is NonNull with any casing'. It's a specific curated list of well known non-null indicating annotations. I've fixed the docs; the doc fix will go out with the next lombok stable release.

2. One could make the case that this curated list needs to be expandable with lombok.config. We're thinking about making this possible.

NB: Expanding the list of annotations which trigger injected null checks into your hand written methods has been discussed and rejected; it's rather impactful to inject invisible code into your hand written methods, so we're going to restrict that functionality solely to `@lombok.NonNull`.

**Comment by janeisklar** [comment]
While your answer obviously isn't what I wanted to hear, I nevertheless would like to thank you for your thorough response!

Furthermore, regarding your consideration to make the annotation configurable I would like to lay out why we want to use our own annotation over the one that comes with lombok: The JSR-305 annotations are readily understood by IntelliJ without having to ask everyone in the team to configure their IDE to also understand the lombok annotation (used to point out potential null-issues in the IDE). The same can be said about code quality tool such as Sonar. (I'm not even sure if Sonar can be configured to pick up `lombok.NonNull`.)

Even if we were to use the lombok's `@NonNull` annotation we would still have to include the JSR-305 dependency as we also use `@Nullable`. Having both `@lombok.NonNull`, `@javax.annotation.Nonnull` and `@javax.annotation.Nullable` to choose from—I think—will make it rather confusing for the team. Asking them to always use `javax.annotation.*` seems a bit more feasible, if only by a small margin.

TL;DR: We would very much welcome it if this annotation was configurable:)
